File: week1/hw1/hw1prob1.py
# (c) 2023 Michael Fitzgerald (mpfitz@ucla.edu)
#
# Using parametric data for a single cycle imported from another program, repeates the
# cycle and adds a scale factor to the parametric curve.  Plots the result.
#
# Demonstrates loading a tabular-formatted text file output from C, plotting a curve,
# annotating a plot with text, and exporting a figure to PDF.
#

import numpy as np
import matplotlib as mpl
import pylab
from astropy.io import ascii

n_cyc = 20 # number of cycles

# load previously generated data from C program
fn = 'hw1prob1.txt' # filename
dat = ascii.read(fn, names=('t','x','y'))

n_pt_cyc = len(dat) # number of points per cycle
n_pt = n_cyc*n_pt_cyc # total number of points

# replicate the x and y arrays n_cyc times, keep in single 1d array
xx = np.concatenate([dat['x']]*n_cyc, axis=0)
yy = np.concatenate([dat['y']]*n_cyc, axis=0)

# scale factor
a_min, a_max = 1., 1.3 # range
a_pow = 2. # power-law index
a = a_min*np.exp(np.linspace(0., 1., n_pt)**a_pow * np.log(a_max/a_min))

# 2d points along parametric curve (apply scale factor)
y = a*yy
x = a*xx

# plot points
fig = pylab.figure(figsize=(5.5,5.5))
ax = fig.add_subplot(111, aspect='equal')
ax.plot(x, y, c='r')

# add annotation
ax.text(-1.5, 1.5, "I")
ax.text(1.25, -1.25, "Astro 142!")

# show/save figure
#pylab.draw()
#pylab.show()
fig.savefig("hw1prob1.pdf")

File: week1/hw1/hw1prob1.c
/*
 *
 * (c) 2023 Michael Fitzgerald (mpfitz@ucla.edu)
 *
 * This creates a text file containing a parametric curve.  Demonstrates dynamic allocation of
 * arrays and exporting values in a tabular formatted text file.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <math.h>

int main(int argc, char* argv[]) {
  static int n_pt_cyc = 100;   // number of points per cycle
  static double tmin = 0.;     // minimum value of t
  static double tmax = 2*M_PI; // maximum value of t

  // create a linearly spaced array of values
  double* t = (double*) malloc(n_pt_cyc * sizeof(double));
  double tstep = (tmax-tmin)/(double)n_pt_cyc; // step size
  if (t == NULL) {
    printf("Failed to allocate memory.\n");
    exit(0);
  }
  for (int i=0; i<n_pt_cyc; i++) {
    t[i] = tmin + (double)i*tstep;
  }

  // compute dependent variables
  double* x = (double*) malloc(n_pt_cyc * sizeof(double));
  double* y = (double*) malloc(n_pt_cyc * sizeof(double));
  for (int i=0; i<n_pt_cyc; i++) {
    x[i] = sin(t[i])*(1-cos(t[i]));
    y[i] = cos(t[i])*(1-cos(t[i])) + 1;
  }
  
  // output values
  static char* fn = "hw1prob1.txt";
  FILE *f = fopen(fn, "w");
  for (int i=0; i<n_pt_cyc; i++) fprintf(f, "%.17g %.17g %.17g\n", t[i], x[i], y[i]);
  fclose(f);
  
  // cleanup
  free(y);
  free(x);
  free(t);
  return 0;
}
